src/profile-logic/profile-data.js

```diff
@@ -223,6 +223,126 @@ export function getCallNodeInfo(
+    let length = 0;
+
+    function addCallNode(
+      funcIndex: number,
+      categoryIndex: number,
+      subcategoryIndex: number,
+      inlinedIntoSymbol: IndexIntoNativeSymbolTable | null
+    ) {
+      if (funcToCallNodeIndex[funcIndex] === -1) {
+        // we found a new function
+        funcToCallNodeIndex[funcIndex] = length;
+        callNodeToFuncIndex.push(funcIndex);
+        length++;
+      }
+      const callNodeIndex = funcToCallNodeIndex[funcIndex]; // index into the other tables
+      prefix[callNodeIndex] = -1; // we have a flat call tree
+      func[callNodeIndex] = funcIndex;
+      depth[callNodeIndex] = 0; // we have a flat call tree
+
+      if (sourceFramesInlinedIntoSymbol[callNodeIndex] === undefined) {
+        // first time we visit it
+        sourceFramesInlinedIntoSymbol[callNodeIndex] = inlinedIntoSymbol;
+        category[callNodeIndex] = categoryIndex;
+        subcategory[callNodeIndex] = subcategoryIndex;
+      } else {
+        if (
+          sourceFramesInlinedIntoSymbol[callNodeIndex] !== inlinedIntoSymbol
+        ) {
+          sourceFramesInlinedIntoSymbol[callNodeIndex] = -1; // conflicting information
+        }
+        if (category[callNodeIndex] !== categoryIndex) {
+          category[callNodeIndex] = defaultCategory;
+        } else {
+          category[callNodeIndex] = categoryIndex;
+        }
+        if (subcategory[callNodeIndex] !== subcategoryIndex) {
+          subcategory[callNodeIndex] = 0;
+        } else {
+          subcategory[callNodeIndex] = subcategoryIndex;
+        }
+      }
+    }
+
+    // Go through each stack, and create a new callNode table, which is based off of
+    // functions rather than frames.
+    for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+      const frameIndex = stackTable.frame[stackIndex];
+      const categoryIndex = stackTable.category[stackIndex];
+      const subcategoryIndex = stackTable.subcategory[stackIndex];
+      const funcIndex = frameTable.func[frameIndex];
+      const inlinedIntoSymbol =
+        frameTable.inlineDepth[frameIndex] > 0
+          ? frameTable.nativeSymbol[frameIndex]
+          : null;
+
+      addCallNode(
+        funcIndex,
+        categoryIndex,
+        subcategoryIndex,
+        inlinedIntoSymbol
+      );
+      stackIndexToCallNodeIndex[stackIndex] = funcToCallNodeIndex[funcIndex]; // map the stackIndex correctly
+    }
+
+    const callNodeTable: CallNodeTable = {
+      prefix: new Int32Array(prefix),
+      func: new Int32Array(func),
+      category: new Int32Array(category),
+      subcategory: new Int32Array(subcategory),
+      depth: new Array(length).fill(0),
+      innerWindowID: new Float64Array(length),
+      sourceFramesInlinedIntoSymbol,
+      length: length,
+    };
+
+    return {
+      callNodeInfo: { callNodeTable, stackIndexToCallNodeIndex },
+      funcToCallNodeIndex,
+      callNodeToFuncIndex: Int32Array.from(callNodeToFuncIndex),
+    };
+  });
+}
```

Comment:
```suggestion
/**
 * Generates a flat call tree for the FunctionTable view, with one call node per
 * function. The retun value consists of a CallNodeInfo which contains the
 * CallNodeTable, and of two maps to convert between an IndexIntoFuncTable and
 * an IndexIntoCallNodeTable.
 */
export function getFunctionTableCallNodeInfo(
  stackTable: StackTable,
  frameTable: FrameTable,
  funcTable: FuncTable,
  defaultCategory: IndexIntoCategoryList
): CallNodeInfoWithFuncMapping {
  return timeCode('getFunctionTableCallNodeInfo', () => {
    const funcToCallNodeIndex = new Int32Array(funcTable.length).fill(-1);
    const callNodeToFuncIndex = [];
    const stackIndexToCallNodeIndex = new Uint32Array(stackTable.length);

    // The callNodeTable components, per function index
    const func: Array<IndexIntoFuncTable> = [];
    const category: Array<IndexIntoCategoryList> = [];
    const subcategory: Array<IndexIntoSubcategoryListForCategory> = [];
    const sourceFramesInlinedIntoSymbol: Array<
      IndexIntoNativeSymbolTable | -1 | null
    > = [];

    function ensureCallNodeForFunction(
      funcIndex: number,
      categoryIndex: number,
      subcategoryIndex: number,
      inlinedIntoSymbol: IndexIntoNativeSymbolTable | null
    ): IndexIntoCallNodeTable {
      let callNodeIndex = funcToCallNodeIndex[funcIndex];
      if (callNodeIndex === -1) {
        // We found a new function.
        callNodeIndex = callNodeToFuncIndex.length;
        callNodeToFuncIndex.push(funcIndex);
        funcToCallNodeIndex[funcIndex] = callNodeIndex;

        func[callNodeIndex] = funcIndex;
        sourceFramesInlinedIntoSymbol[callNodeIndex] = inlinedIntoSymbol;
        category[callNodeIndex] = categoryIndex;
        subcategory[callNodeIndex] = subcategoryIndex;
      } else {
        // Combine sourceFramesInlinedIntoSymbol, category and subcategory info.
        if (
          sourceFramesInlinedIntoSymbol[callNodeIndex] !== inlinedIntoSymbol
        ) {
          sourceFramesInlinedIntoSymbol[callNodeIndex] = -1; // conflicting information
        }
        if (category[callNodeIndex] !== categoryIndex) {
          category[callNodeIndex] = defaultCategory;
        } else {
          category[callNodeIndex] = categoryIndex;
        }
        if (subcategory[callNodeIndex] !== subcategoryIndex) {
          subcategory[callNodeIndex] = 0;
        } else {
          subcategory[callNodeIndex] = subcategoryIndex;
        }
      }
      return callNodeIndex;
    }

    // Go through each stack, and create a new callNode table, which is based off of
    // functions rather than frames.
    for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
      const frameIndex = stackTable.frame[stackIndex];
      const categoryIndex = stackTable.category[stackIndex];
      const subcategoryIndex = stackTable.subcategory[stackIndex];
      const funcIndex = frameTable.func[frameIndex];
      const inlinedIntoSymbol =
        frameTable.inlineDepth[frameIndex] > 0
          ? frameTable.nativeSymbol[frameIndex]
          : null;

      const callNodeIndex = ensureCallNodeForFunction(
        funcIndex,
        categoryIndex,
        subcategoryIndex,
        inlinedIntoSymbol
      );
      stackIndexToCallNodeIndex[stackIndex] = callNodeIndex;
    }

    const length = callNodeToFuncIndex.length;
    const callNodeTable: CallNodeTable = {
      prefix: new Int32Array(length).fill(-1),
      func: new Int32Array(func),
      category: new Int32Array(category),
      subcategory: new Int32Array(subcategory),
      depth: new Array(length).fill(0),
      innerWindowID: new Float64Array(length),
      sourceFramesInlinedIntoSymbol,
      length: length,
    };

    return {
      callNodeInfo: { callNodeTable, stackIndexToCallNodeIndex },
      funcToCallNodeIndex,
      callNodeToFuncIndex: Int32Array.from(callNodeToFuncIndex),
    };
  });
}
```